io.openems.edge.ess.mr.gridcon/src/io/openems/edge/ess/mr/gridcon/GridconPcsImpl.java

```diff
@@ -0,0 +1,1698 @@
+		IntegerReadChannel errorCountChannel = this.channel(GridConChannelId.CCU_ERROR_COUNT);
+		int errorCount = errorCountChannel.value().orElse(-1);
+		return "Gridcon CCU state: " + state + "; Error count: " + errorCount + "; Active Power: "
+				+ this.getActivePower();
+
+	}
+
+	// private String getGridMeasurements() {
+	// StringBuilder s = new StringBuilder();
+	//
+	// s.append("Current L1: ");
+	// s.append(getCurrentL1Grid());
+	// s.append("\n");
+	//
+	// s.append("Current L2: ");
+	// s.append(getCurrentL2Grid());
+	// s.append("\n");
+	//
+	// s.append("Current L3: ");
+	// s.append(getCurrentL3Grid());
+	// s.append("\n");
+	//
+	// s.append("Current N: ");
+	// s.append(getCurrentLNGrid());
+	// s.append("\n");
+	//
+	// s.append("Active Power L1: ");
+	// s.append(getActivePowerL1Grid());
+	// s.append("\n");
+	//
+	// s.append("Active Power L2: ");
+	// s.append(getActivePowerL2Grid());
+	// s.append("\n");
+	//
+	// s.append("Active Power L3: ");
+	// s.append(getActivePowerL3Grid());
+	// s.append("\n");
+	//
+	// s.append("Active Power Sum: ");
+	// s.append(getActivePowerSumGrid());
+	// s.append("\n");
+	//
+	// s.append("Reactive Power L1: ");
+	// s.append(getReactivePowerL1Grid());
+	// s.append("\n");
+	//
+	// s.append("Reactive Power L2: ");
+	// s.append(getReactivePowerL2Grid());
+	// s.append("\n");
+	//
+	// s.append("Reactive Power L3: ");
+	// s.append(getReactivePowerL3Grid());
+	// s.append("\n");
+	//
+	// s.append("Reactive Power Sum: ");
+	// s.append(getReactivePowerSumGrid());
+	// s.append("\n");
+	//
+	// s.append("Apparent Power L1: ");
+	// s.append(getApparentPowerL1Grid());
+	// s.append("\n");
+	//
+	// s.append("Apparent Power L2: ");
+	// s.append(getApparentPowerL2Grid());
+	// s.append("\n");
+	//
+	// s.append("Apparent Power L3: ");
+	// s.append(getApparentPowerL3Grid());
+	// s.append("\n");
+	//
+	// s.append("Apparent Power Sum: ");
+	// s.append(getApparentPowerSumGrid());
+	// s.append("\n");
+	//
+	//
+	// return s.toString();
+	// }
+
+	@Override
+	public void setPower(int activePower, int reactivePower) {
+		float maxApparentPower = this.getMaxApparentPower();
+		/*
+		 * !! signum, MR calculates negative values as discharge, positive as charge.
+		 * Gridcon sets the (dis)charge according to a percentage of the
+		 * MAX_APPARENT_POWER. So 0.1 => 10% of max power. Values should never take
+		 * values lower than -1 or higher than 1.
+		 */
+
+		this.activePowerPreset = activePower;
+
+		float activePowerFactor = (-1) * activePower / maxApparentPower;
+		float reactivePowerFactor = (-1) * reactivePower / maxApparentPower;
+
+		this.commands.setParameterPref(activePowerFactor);
+		this.commands.setParameterQref(reactivePowerFactor);
+	}
+
+	protected void writeCommands() throws IllegalArgumentException, OpenemsNamedException {
+
+		Commands c = this.commands;
+
+		this.writeValueToChannel(GridConChannelId.COMMAND_CONTROL_WORD_PLAY, c.getPlayBit());
```

Comment:
`writeValueToChannel` ist so eine typische Methode, die nur ein Wrapper um die Channel-Methode `setNextValue()` ist. Der Nachteil ist, dass diese Wrapper oft Internas verstecken, die ich bewusst machen möchte (z. B. explizites Handling von 'null'-Werten, die immer vorkommen können) oder den Quellcode für andere OpenEMS-Entwickler, die sich an die "gängigen OpenEMS-Paradigmen" halten, schwer lesbar machen.

Auch wenn der Ansatz, den ich jetzt immer umsetze mit `_setXYZ()`-Methoden nicht weniger Quellcode bedeutet, so finde ich ihn doch in der Anwendung ziemlich elegant. Und konsequent überall umgesetzt ist er auch sehr gut lesbar und verständlich. Beispiel: https://github.com/OpenEMS/openems/blob/develop/io.openems.edge.controller.api/src/io/openems/edge/controller/api/Controller.java#L65